bundles/org.openhab.binding.meater/src/main/java/org/openhab/binding/meater/internal/api/MeaterRestAPI.java

```diff
@@ -0,0 +1,190 @@
+import org.eclipse.jetty.client.HttpClient;
+import org.eclipse.jetty.client.HttpResponseException;
+import org.eclipse.jetty.client.api.ContentResponse;
+import org.eclipse.jetty.client.api.Request;
+import org.eclipse.jetty.client.api.Response;
+import org.eclipse.jetty.client.util.StringContentProvider;
+import org.eclipse.jetty.http.HttpHeader;
+import org.eclipse.jetty.http.HttpMethod;
+import org.eclipse.jetty.http.HttpStatus;
+import org.openhab.binding.meater.internal.MeaterBridgeConfiguration;
+import org.openhab.binding.meater.internal.MeaterException;
+import org.openhab.binding.meater.internal.dto.MeaterProbeDTO;
+import org.openhab.binding.meater.internal.dto.MeaterProbeDTO.Device;
+import org.openhab.core.i18n.LocaleProvider;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.Gson;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParseException;
+import com.google.gson.JsonParser;
+import com.google.gson.JsonSyntaxException;
+
+/**
+ * The {@link MeaterRestAPI} class defines the Meater REST API
+ *
+ * @author Jan Gustafsson - Initial contribution
+ */
+@NonNullByDefault
+public class MeaterRestAPI {
+    private static final String API_ENDPOINT = "https://public-api.cloud.meater.com/v1/";
+    private static final String JSON_CONTENT_TYPE = "application/json";
+    private static final String LOGIN = "login";
+    private static final String DEVICES = "devices";
+    private static final int MAX_RETRIES = 3;
+
+    private final Logger logger = LoggerFactory.getLogger(MeaterRestAPI.class);
+    private final Gson gson;
+    private final HttpClient httpClient;
+    private final MeaterBridgeConfiguration configuration;
+    private String authToken = "";
+    private String userId = "";
+    private LocaleProvider localeProvider;
+
+    public MeaterRestAPI(MeaterBridgeConfiguration configuration, Gson gson, HttpClient httpClient,
+            LocaleProvider localeProvider) {
+        this.gson = gson;
+        this.configuration = configuration;
+        this.httpClient = httpClient;
+        this.localeProvider = localeProvider;
+    }
+
+    public boolean refresh(Map<String, MeaterProbeDTO.Device> meaterProbeThings) {
+        try {
+            MeaterProbeDTO dto = getDevices(MeaterProbeDTO.class);
+            List<Device> devices = dto.getData().getDevices();
+
+            if (devices != null) {
+                if (!devices.isEmpty()) {
+                    for (Device meaterProbe : devices) {
+                        meaterProbeThings.put(meaterProbe.id, meaterProbe);
+                    }
+                } else {
+                    meaterProbeThings.clear();
+                }
+            }
+            return true;
+        } catch (MeaterException | JsonSyntaxException e) {
+            logger.warn("Failed to refresh! {}", e.getMessage());
+        }
+        return false;
+    }
+
+    public void login() throws MeaterException {
```

Comment:
This could be changed into a private method to be called when:
- `authToken` is invalid (not set). This will happen the first time after the binding is started.
- `getFromApi()` has returned HTTP code 401. This will happen when JWT is reset.

This would eliminate all authentication logic from the handler.

This is quite similar to the Bosch Indego API, except the token actually has an expiration time in this case. You could have a look at:
https://github.com/openhab/openhab-addons/blob/main/bundles/org.openhab.binding.boschindego/src/main/java/org/openhab/binding/boschindego/internal/IndegoController.java

Reply:
Fixed!